components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/device/codegenerator/GenerateKeys.java

```diff
@@ -40,7 +47,42 @@ public GenerateKeys() {
      */
     public static String getKey(int num) {
 
-        return RandomStringUtils.random(num, 0, Constants.KEY_SET.length(), false, false,
-                Constants.KEY_SET.toCharArray(), new SecureRandom());
+        int userCodeLength;
+        int keysetEnd;
+        char[] subKeyset;
+        int configuredUserCodeLength;
+        String configuredKeySet = IdentityUtil.getProperty(Constants.CONF_KEY_SET);
+        String configuredLength = IdentityUtil.getProperty(Constants.CONF_USER_CODE_LENGTH);
+        try {
+            configuredUserCodeLength = (StringUtils.isNotBlank(configuredLength) ? Integer.parseInt(configuredLength) :
+                    Constants.KEY_LENGTH);
+        } catch (NumberFormatException e) {
+            log.error("Error while converting user_code length " + configuredLength + " to integer. ", e);
+            configuredUserCodeLength = Constants.KEY_LENGTH;
+        }
+        userCodeLength = Math.max(configuredUserCodeLength, num);
+        if (log.isDebugEnabled()) {
+            log.debug("User defined keyset : " + configuredKeySet + " and user_code length : " +
+                    configuredUserCodeLength);
+        }
+        if (StringUtils.isNotBlank(configuredKeySet)) {
+            keysetEnd = configuredKeySet.length();
+            subKeyset = configuredKeySet.toCharArray();
+        } else {
+            keysetEnd = Constants.KEY_SET.length();
+            subKeyset = Constants.KEY_SET.toCharArray();
+        }
+        return RandomStringUtils.random(userCodeLength, 0, keysetEnd, false, false,
+                subKeyset, new SecureRandom());
+    }
+
+    /**
+     * This method is used to generate current quantifier.
+     *
+     * @return Current quantized time period user_code belongs.
+     */
+    public static long getCurrentQuantifier() {
+
+        return (Instant.now().getEpochSecond() / (2 * DeviceFlowUtil.getConfiguredExpiryTime()));
```

Comment:
Shall we explain the logic here? Why 2?

Reply:
https://github.com/wso2/product-is/issues/7348#issuecomment-593761350